shared/src/commonMain/kotlin/io/customer/shared/tracking/api/model/Extensions.kt

```diff
@@ -0,0 +1,93 @@
+package io.customer.shared.tracking.api.model
+
+import io.customer.shared.sdk.meta.IdentityType
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.tracking.model.Activity
+import io.customer.shared.tracking.model.TrackingError
+import io.customer.shared.tracking.model.type
+
+internal val TrackingError.taskStatus: QueueTaskStatus
+    get() = if (reason.isFixable) QueueTaskStatus.FAILED
+    else QueueTaskStatus.INVALID
+
+internal val TrackingErrorReason.isFixable: Boolean
+    get() = when (this) {
+        TrackingErrorReason.REQUIRED -> true
+        TrackingErrorReason.INVALID,
+        TrackingErrorReason.PARSE_ERROR,
+        -> false
+    }
+
+internal val BatchTrackingResponse.isSuccessful: Boolean
+    get() = statusCode in 200..299
+
+internal val BatchTrackingResponse.isServerUnavailable: Boolean
+    get() = statusCode >= 500
+
+internal fun Activity.toTrackingRequest(
+    identityType: IdentityType,
+    profileIdentifier: String,
+) = when (this) {
+    is Activity.IdentifyProfile -> TrackingRequest(
+        identityType = identityType,
+        profileIdentifier = profileIdentifier,
+    )
+    is Activity.AddDevice -> TrackingRequest(
+        identityType = identityType,
+        profileIdentifier = profileIdentifier,
+        device = device,
+    )
+    is Activity.DeleteDevice -> TrackingRequest(
+        identityType = identityType,
+        profileIdentifier = profileIdentifier,
+        device = device,
+    )
+    is Activity.Metric -> TrackingRequest(
+        identityType = identityType,
+        profileIdentifier = profileIdentifier,
+        metric = Metric(
+            deliveryID = deliveryId,
+            deviceToken = deviceToken,
+            event = metricEvent,
+            timestamp = timestamp,
+        ),
+    )
+    is Activity.Page -> TrackingRequest(
+        identityType = identityType,
+        profileIdentifier = profileIdentifier,
+        name = name,
+    )
+    is Activity.Screen -> TrackingRequest(
+        identityType = identityType,
+        profileIdentifier = profileIdentifier,
+        name = name,
+    )
+    is Activity.Event -> TrackingRequest(
+        identityType = identityType,
+        profileIdentifier = profileIdentifier,
+        name = name,
+    )
+}
+
+private val IdentityType.key: String
+    get() = when (this) {
+        IdentityType.CIO_ID -> "cio_id"
+        IdentityType.ID -> "id"
+        IdentityType.EMAIL -> "email"
+    }
```

Comment:
This code has important logic in it that I believe should be moved out of this file and into the file of the appropriate data type. Encapsulate the logic into 1 file instead of fragment it here. 

I don't mind 1 file that contains a random collection of extension functions as long as those functions are convenience functions such as:

```kotlin
internal val BatchTrackingResponse.isSuccessful: Boolean
    get() = statusCode in 200..299
```

Reply:
Sorry I couldn't get you completely here. Are you suggesting to create separate file like `ActivityExt.kt`?

Just for clarity, the code here just transformes `Activity` into our network request model `TrackingRequest`.

Reply:
I am suggesting that instead of having 1 file called `Extensions` that contain extensions of multiple different data types, move extensions code into the file where the data type is defined. 

For example, move `val IdentityType.key: String` into the file where `IdentityType` is defined. And move `fun Activity.toTrackingRequest` into the file where `Activity` is defined. 

There are some scenarios where we are adding extensions to a data type our code base doesn't define. When using dependencies we do that. In those scenarios, I can see creating a file, `CoroutinesExtensions` to be OK to do.

Reply:
I understand the point now. But these method are only required by network layer and I wanted to keep them separate. e.g. `Activity` class is shared between all layers (i.e. Queue, Database and Network). But `TrackingRequest` is only used by network layer. Can you think of something which prevents exposing these methods to other layers and addresses your concern? I want to restrict these methods to network layer only (at least at package level).

Reply:
Ah, I didn't think of that use case. 

What about having `fun Activity.toTrackingRequest` exist in the `TrackingRequest` file? 
`val IdentityType.key: String` exist in the `IdentityType` file?

Reply:
This sounds good, I'm moving it to `TrackingRequest` file.

`IdentityType.key` is also only required by `TrackingRequest`. So we can either put this too to `TrackingRequest` file, or may be just remove the extension and evaluate it using `when` in `fun Activity.toTrackingRequest` 🤔

Reply:
Pushed the changes, added `IdentityType.key` to `fun Activity.toTrackingRequest`